src/Security.php

```diff
@@ -24,296 +26,50 @@
  * @author Derek Jones <derek.jones@ellislab.com>
  * @author Graham Campbell <graham@alt-three.com>
  */
-class Security
+final class Security
 {
     /**
-     * The evil attributes.
-     *
-     * @var string[]
+     * @var AntiXSS
      */
-    protected $evil;
+    private $antiXss;
 
     /**
      * Create a new security instance.
      *
      * @param string[]|null $evil
+     * @param string        $replacement
+     * @param AntiXSS|null  $antiXss
      *
      * @return void
      */
-    public function __construct(array $evil = null)
+    public function __construct(array $evil = null, string $replacement = '', AntiXSS $antiXss = null)
     {
-        $this->evil = $evil ?: ['(?<!\w)on\w*', 'style', 'xmlns', 'formaction', 'form', 'xlink:href', 'FSCommand', 'seekSegmentTime'];
-    }
-
-    /**
-     * XSS clean.
-     *
-     * @param string|string[] $str
-     *
-     * @return string|string[]
-     */
-    public function clean($str)
-    {
-        if (is_array($str)) {
-            foreach ($str as $key => &$value) {
-                $str[$key] = $this->clean($value);
+        $evilRegex = [];
+        if ($evil && $evil !== []) {
+            foreach ($evil as $regex) {
+                $evilRegex[$regex] = $replacement;
             }
-
-            return $str;
         }
 
-        $i = 0;
-        $hash = bin2hex(random_bytes(20));
-
-        do {
-            $i++;
-            $processed = $this->process($str, $hash);
-        } while ($i < 3 && $processed !== $str);
-
-        return $processed;
-    }
-
-    /**
-     * Process a string for cleaning.
-     *
-     * @param string $str
-     * @param string $hash
-     *
-     * @return string
-     */
-    protected function process(string $str, string $hash)
-    {
-        $str = self::removeInvisibleCharacters($str);
-
-        if (stripos($str, '%') !== false) {
-            do {
-                $original = $str;
-                $str = preg_replace_callback(
-                    '#%(?:\s*[0-9a-f]){2,}#i',
-                    [$this, 'urlDecodeSpaces'],
-                    rawurldecode($str)
-                );
-            } while ($original !== $str);
-
-            unset($original);
+        if ($antiXss === null) {
```

Comment:
This would be neater in a single line I think?

Reply:
yep, done